File: Graph/207_CourseSchedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adjacent = collections.defaultdict(list)
        indegree = [0] * numCourses

        for pre in prerequisites:
            adjacent[pre[1]].append(pre[0])
            indegree[pre[0]] += 1

        queue = []
        for i in range(numCourses):
            if indegree[i] == 0:
                queue.append(i)

        visited = 0
        while queue:
            node = queue.pop(0)
            visited += 1
            for adj in adjacent[node]:
                indegree[adj] -= 1
                if indegree[adj] == 0:
                    queue.append(adj)

        return visited == numCourses

File: Java Solutions/207_CourseSchedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> adjacent = new ArrayList<>();
        int[] indegree = new int[numCourses];
        for (int i = 0; i < numCourses; i++) {
            adjacent.add(new ArrayList<>());
        }
        for (int[] pre : prerequisites) {
            adjacent.get(pre[1]).add(pre[0]);
            indegree[pre[0]]++;
        }

        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                queue.offer(i);
            }
        }

        int visited = 0;
        while (!queue.isEmpty()) {
            int node = queue.poll();
            visited++;
            for (int adj : adjacent.get(node)) {
                indegree[adj]--;
                if (indegree[adj] == 0) {
                    queue.offer(adj);
                }
            }
        }

        return visited == numCourses;
    }
}